web/js/codeworld_shared.js

```diff
@@ -2004,15 +1836,337 @@ function updateChildrenIndexes(node) {
 }
 
 function updateTreeOnNewProjectCreation() {
-  const selectedNode = getSelectedNode();
+  const selectedNode = DirTree.getSelectedNode();
 
-  if (selectedNode && utils.directoryTree.isProject(selectedNode)) {
-    utils.directoryTree.clearSelectedNode();
+  if (selectedNode && DirTree.isProject(selectedNode)) {
+    DirTree.clearSelectedNode();
 
     // Select parent folder (if any) as new project's location is
     // determined by the selection in the directory tree.
     if (pathToRootDir(selectedNode)) {
-      utils.directoryTree.selectNode(selectedNode.parent);
+      DirTree.selectNode(selectedNode.parent);
+    }
+  }
+}
+
+function clearWorkspace() {
+  DirTree.clearSelectedNode();
+
+  setCode('');
+}
+
+function isEditorClean() {
+  if (!window.codeworldEditor) {
+    return true;
+  }
+
+  const doc = window.codeworldEditor.getDoc();
+
+  if (window.savedGeneration === null) return doc.getValue() === '';
+  else return doc.isClean(window.savedGeneration);
+}
+
+/*
+ * Updates all UI components to reflect the current state.  The general pattern
+ * is to modify the state stored in variables and such, and then call updateUI
+ * to get the visual presentation to match.
+ */
+function updateUI() {
```

Comment:
This definitely feels like the wrong place for this function.  The point of this function is to adjust visibility of specific elements on the page.  `codeworld_shared.js` is the code that's shared by multiple different pages (blocks.html and env.html), which have different elements on them.  And it looks like blocks still has its own `updateUI` as well?

What about this?  Instead of having one updateUI function, have codeworld_shared.js export a function to add an event handler, and fire events instead of calling updateUI directly.  Then each environment can provide its own updateUI function that does the right things for that page.